🚀 2단계 - HTTP 서버 구현하기

jivebreaddev opened the pull request:
## 2단계 요구사항 정리
![image](https://github.com/user-attachments/assets/91e5cde8-2ab5-47d6-80bb-ea189e6e97d4)
![image](https://github.com/user-attachments/assets/4d49e038-591b-4f20-ba90-956179e17ead)

## 1단계 PR 입니다. 
https://github.com/next-step/java-http/pull/61
(코멘트 달아 뒀는데 필요하신 궁금하신 사항은 추가적으로 말씀주세요!)

## 2단계 요구사항 정리
- [x] 리뷰 반영 
  - [x] Response에 미구현된 것 확인
  - [x] Review 질문 답변
  - [x] GreetingController에 'index.html'에 suffix가 있어야 하는 이유
  - [x] 동시성 문제 코멘트에 대한 정리
  - [x] 상수 네이밍 확인 DELIMITER
  - [x] STREAM 닫지 못한것 확인
  - [x] String.SPLIT PATTERN 객체 생성 주입
  - [x] 코드간 스타일 (상수 + 변수 개행)
- [x] 요구사항 1 - GET /index.html 응답하기
  - [x] Resource를 반환하는 기능 지원
- [x] 요구사항 2 - GET CSS 지원하기
  - [x] Request에 header 추가로 읽어오기 (Content-Type 추가)
  - [x] Response에 header 추가하기 (Content-Type 추가)
- [x] 요구사항 3 - GET QUERY 파싱하기
  - [x] InMemoryUserRepository 사용해서 회원 조회
- [x] 요구사항 4 - 공통 설계 요구사항
  - [x] HTTP RESPONSE를 생성하는 책임을 분리 (Factory)
  - [x] HTTP RESPONSE 도메인 객체 생성 + HEADER 추가
  - [x] HTTP Request에 따른 Resource 맵핑 하는 책임을 가진 ControllerFactoryProvider 생성
  - [x] ControllerFactoryProvider가 url 맵핑 리소스들을 스캐닝해와서 HttpResponse를 만드는 Factory와 연결한다. 
  - [x] Factory가 RequestMethod에 따라 Strategy를 구성해서 Response를 반환합니다.
  - [x] Response Test 생성
  - [x] Request Test 생성

## 질문 및 세부구현 정리
- Step4 에서 Controller 인터페이스를 구현해야하는 것으로 보입니다.
- doGet과 doPost가 객체지향적으로 보이지 않아 전략 패턴으로 합성을 사용했습니다. 
- 톰캣은 왜 이런 방식으로 구현을 했을까요 흠... Servlet 스펙을 지키기 위한것이겠죠?

### 세부 구현 
- Response 객체 구현
- Response 객체를 생성하기 위한 생성 책임을 Factory가 가집니다.
- Controller 하위 클래스들이 HttpRequest에 따른 HttpResponse를 생성하는 Factory에 맵핑하는 역할을 합니다.
- Factory 하위에 전략 패턴을 도입해서 합성을 통해 맵핑될 수 있도록 변경하였습니다.



### 학습한 부분들

- Spring 에서 사용하는 Controller가 왜 나온지 Mapping을 하면서 이해했습니다.
- Spring에서 Resource와 @Controller Annotation들을 스캐닝해서 맵핑부분들을 생성하는것에 대한 이해를 했습니다.
- Http Response의 구조
- Http Response를 생성하기 위한 로직 플로우를 이해했습니다.

jivebreaddev committed:
DOCS(README): 요구사항 정리 및 피드백 반영하였습니다.

jivebreaddev committed:
Style(README): 요구사항 정리 및 피드백 반영하였습니다.
- split 반영
- HttpFactoryProvider에서 HttpRequestFactoryProvider로 명명 변경

jivebreaddev committed:
Test(Request): Request 도메인의 객체들에 대한 테스트를 추가 하였습니다.

jivebreaddev committed:
Test(Request): Request 도메인의 객체들에 대한 테스트를 추가 하였습니다.

jivebreaddev committed:
Feature(Response): Response 도메인의 객체들에 기능을 추가하였습니다.
- HttpResponse
   - HttpResponseProtocol
   - HttpResponseVersion
   - StatusLine
      - StatusCode
- HttpResponse의 생성을 담당하는 Builder

jivebreaddev committed:
Test(Response): Response 도메인의 객체들에 기능을 테스트 하였습니다.
- HttpResponse
   - HttpResponseProtocol
   - HttpResponseVersion
   - StatusLine
      - StatusCode
- HttpResponse의 생성을 담당하는 Builder

jivebreaddev committed:
feat(Controller): Http 요청에 따른 알맞은 HttpResponse 생성에 대한 Factory를 맵핑 역할하는 기능을 추가 하였습니다.
- ControllerProvider
- ControllerFactory
  - 전략 패턴을 사용한 RequestMethodStrategy

jivebreaddev committed:
feat(HttpProcessor): 해당 요구사항들을 모두 충족시키는지 Test와 생성된 인터페이스를 Processor에서 사용하였습니다.

jivebreaddev committed:
feat(ControllerFactoryProvider): 코멘트 제거

catsbi commented on README.md:
```diff
@@ -50,7 +50,33 @@
-  - [] HTTP RESPONSE를 생성하는 책임을 분리
   - [x] 커스텀 예외 생성
+
+## 2단계 요구사항 정리
```
꼼꼼한 readme.md 💯 
개인적으로, 기능 개발에 지장을 주거나 당장 문제는 없으니 이런 문서화를 안하는 경우가 많지만 저는 여유가 조금이라도 된다면 readme.md를 꼼꼼히 작성하는 습관을 키우는걸 권장합니다. 사람의 기억력은 단기던 장기던 결국 한계가 있다보니, 차후 유지보수 측면에서 문서화가 제대로 되어있지 않으면, 문제가 되더라구요. 

그리고 개발도중에도 진행도 파악등은 지라 이슈를 통해서도 가능하지만 readme.md를 통해서도 하는걸 좋아합니다 ㅎㅎ

jivebreaddev replied:
넵넵! 습관이 되게 중요하더라구요!

readme를 작성하면서 배운 부분들도 다시 복습하고 과거에 다시 찾는데 시간이 안걸려서 좋았던것같아요!!

언급해주셔서 감사합니다!

catsbi commented on study/src/test/java/study/IOStreamTest.java:
```diff
@@ -239,4 +240,48 @@ class InputStreamReader_학습_테스트 {
+//
+//        actual.append(bufferedReader.readLine()).append("\r\n");
+//            assertThatThrownBy(() -> bufferedReader.readLine())
+//                .isInstanceOf(RuntimeException.class);
```
형상관리도구(git, svn)가 있기 때문에 사용하지 않는 코드는 제거해주세요. 제거하지 말아야 할 이유가 있다면 TODO 주석을 통해 기록을 남기는걸 추천드려요.

jivebreaddev replied:
모두 제거하였습니다.

catsbi commented on study/src/test/java/study/IOStreamTest.java:
```diff
@@ -239,4 +240,48 @@ class InputStreamReader_학습_테스트 {
         }
     }
 
+    /**
+     * BufferedReader가_닫히면_InputStream이_닫힌다
+     */
```
DisplayName으로 작성하셔도 될 것 같네요!

jivebreaddev replied:
변경하였습니다.

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -33,28 +38,24 @@ public void run() {
+                httpRequest.getRequestUrl());
+            HttpResponse httpResponse = controllerFactory.serve(httpRequest);
 
+            log.error(httpResponse.toString());
```
에러가 발생하지 않았는데 에러로그를 찍어야 할까요?

jivebreaddev replied:
log.info로 변경하였습니다.

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/Http11Processor.java:
```diff
@@ -33,28 +38,24 @@ public void run() {
-                responseBody);
+            ControllerFactory controllerFactory = controllerFactoryProvider.provideFactory(
+                httpRequest.getRequestUrl());
+            HttpResponse httpResponse = controllerFactory.serve(httpRequest);
```
httpResponse가 404인경우 null을 반환시키고 있어요. 그러다보니 밑에서 write()를 그대로 호출하면 NPE가 발생할 것 같습니다.

jivebreaddev replied:
null일때, 404.html 을 반환하게끔 변경하였습니다.

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/controller/Controller404Factory.java:
```diff
@@ -0,0 +1,12 @@
+
+    @Override
+    public HttpResponse serve(HttpRequest httpRequest) {
+        return null;
```
Java는 언어레벨에서 Null Safety를 지원하지 않기 때문에 명시적으로 null을반환하면 체크해야할 범위가 너무 많아집니다.

jivebreaddev replied:
해당 404도 지원하도록 변경하였습니다.

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/controller/ControllerDefaultFactory.java:
```diff
@@ -0,0 +1,23 @@
+    public HttpResponse serve(HttpRequest httpRequest) {
+        String message = "Hello world!";
+        return new Http11Response.HttpResponseBuilder()
+            .statusLine("1.1", "OK")
+            .responseHeader("text/html", message.getBytes().length)
```
version과 type, content-type등이 모두 하드코딩되어있는데 열거타입으로 만들어서 관리해보면 어떨까요

jivebreaddev replied:
enum type 모두 지원하게 변경하였습니다!

확실히 깔끔해진것같아요!

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/controller/ControllerFactoryProvider.java:
```diff
@@ -0,0 +1,61 @@
+
+    public ControllerFactoryProvider() {
+        ControllerFactory resourceFactory = new ControllerResourceFactory();
+        staticResources.put("/register.html", resourceFactory);
+        staticResources.put("/login.html", resourceFactory);
+        staticResources.put("/index.html", resourceFactory);
+        staticResources.put("/css/styles.css", resourceFactory);
+        staticResources.put("/img/chart-area.js", resourceFactory);
+        staticResources.put("/img/chart-bar.js", resourceFactory);
+        staticResources.put("/img/chart-pie.js", resourceFactory);
```
모든 정적 리소스를 이렇게 하나하나 추가해줄수는 없습니다. 스프링은 보통 resources 내의 static 디렉토리 접근시 정적 리소스 접근으로 파악하여 허용해주는 방식을 사용하는데, 이를 참고해보면 어떨까요?

jivebreaddev replied:
regex를 통해서 정적리소스라면 맵핑하게끔 변경하였습니다.

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/controller/ControllerResourceFactory.java:
```diff
@@ -0,0 +1,63 @@
+            url = DEFAULT_URL;
+        }
+
+        try {
+            final URL resource = getClass().getClassLoader()
+                .getResource(url);
+            final File file = new File(Objects.requireNonNull(resource).getFile());
+            final Path path = file.toPath();
+            final byte[] content = Files.readAllBytes(path);
+            final String extension =
+                extensionMapper.keySet()
+                    .stream()
+                    .filter(ext -> file.getName().endsWith(ext))
+                    .findAny()
+                    .map(extensionMapper::get)
+                    .orElseGet(() -> "*/*");
+            return new Http11Response.HttpResponseBuilder()
+                .responseHeader(extension, content.length)
+                .statusLine(httpRequest.getVersion(), "OK")
+                .messageBody(content)
+                .build();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
- readAllBytes 함수는 OOM 이슈가 있을 것 같아요 🤔 
- ContentType에 대해서 작성되는 로직이 있으면 좋을 것 같네요. ContentType을 열거타입등으로 만들면 어떨까요? 

```java
public enum ContentType {
    JS("application/javascript;charset=utf-8", "./static"),
    HTML("text/html;charset=utf-8", "./templates"),
    CSS("text/css", "./static"),
    SVG("image/svg+xml", "./static"),
    TTF("application/octet-stream", "./static"),
    OTF("application/x-font-opentype", "./static"),
    WOFF("application/font-woff", "./static"),
    WOFF2("application/font-woff2", "./static"),
    ICO("image/avif", "./templates"),
    JSON("application/json", "");
}
```

이런식으로 말이죠 ㅎㅎ

jivebreaddev replied:
readAllBytes 함수는 OOM 이슈가 있을 것 같아요 🤔


해당 피드백에 대해 변경 방향을 잡고있는데 라이브러리를 사용하지 않게 되면 검증하기까지 마음이 조금 불편해질것같아요!!

혹시 어떤 방식으로 수정해야할지 피드백을 받을 수 있을까요??

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/controller/strategy/NotFoundStrategy.java:
```diff
@@ -0,0 +1,17 @@
+
+    @Override
+    public HttpResponse serve(HttpRequest httpRequest) {
+        return null;
```
null만을 반환하는 함수는 고민해보면 좋을 것 같아요.

jivebreaddev replied:
해당 404 컨트롤러 구현하였습니다.

catsbi reviewed (changes requested):
시창님 2단계 구현 역시 엄청 열심히 해주셨네요 👏 
개발항 부분들에 대해서 체크리스트로 꼼꼼히 체크하면서 개발하신것 같아요. 

그런데 현재 프로젝트를 실행하면 정상적으료 요구사항 페이지 접속이 안되는 이슈가 있는 것 같아요. 
그리고 추가적으로 고민해보면 좋을부분 코멘트 남겨드렸으니 확인 후 다시 리뷰요청 부탁드려요!

jivebreaddev committed:
feat(HttpRequestParser, Processor): inputStream을 닫으면서, OutputStream 닫히는 이슈 수정하였습니다.
- InputStream을 인수로 내리고 BufferReader를 생성하지 않습니다.
- TRY-WITH-RESOURCES에서 바로 생성해서 인수로 전달합니다.

jivebreaddev committed:
test(IOStreamTest): IOStreamTest 주석 삭제 및 구체적인 DisplayName을 작성하였습니다.

jivebreaddev committed:
feature(static): static resources에 대해서 regex를 사용해서 path를 허용했습니다.

jivebreaddev committed:
feature(HttpResponse): ContentType, ProtocolVersion, StatusCode Enum화 이후 사용하였습니다.

jivebreaddev committed:
feature(HttpResponse): ContentType, ProtocolVersion, StatusCode Enum화 이후 사용하였습니다.

jivebreaddev committed:
feature(NOT FOUND): Null 표기 대신 404 표기로 NULL 값을 반환하는 값들 제거 하였습니다.

jivebreaddev committed:
feature(NOT FOUND): Null 표기 대신 404 표기로 NULL 값을 반환하는 값들 제거 하였습니다.

jivebreaddev committed:
feature(NOT FOUND): Null Strategy 삭제

jivebreaddev committed:
test(fix): Test 통과하게 수정하였습니다.

catsbi commented on tomcat/src/main/java/org/apache/coyote/http11/controller/Controller404Factory.java:
```diff
@@ -1,12 +1,40 @@
+                .getResource(NOTFOUND);
+            final File file = new File(Objects.requireNonNull(resource).getFile());
+            final Path path = file.toPath();
+            final byte[] content = Files.readAllBytes(path);
```
🤔  File에 대해서는 마지막 write시점까지는 metadata만 가지고 있고(File) 마지막에 실제로 작성하는 시점에 chuck단위로만 읽어서 쓰도록해보면 어떨까요? length도 모든 content를 바이트로 힙에 올려서 체크하기보단 file.length()를 사용하면 UnixFileSystem에서  native 코드를 쓰니 성능적으로 좀 더 이점이 있을꺼에요.

catsbi reviewed (approved):
시창님 빠르게 피드백 반영해주셨네요 👍 
더 이상 크게 문제가 될만한 부분들은 보이지 않아 바로 머지하겠습니다. 다음단계 진행해주세요!
